Thread stats API

deepikabhavnani opened the pull request:
### Description

Addition of API to get information of all active threads in Mbed OS.

### Pull request type

    [ ] Fix
    [ ] Refactor
    [ ] New target
    [X] Feature
    [ ] Breaking change

CC @SenRamakri 

Dependent on https://github.com/ARMmbed/mbed-os/pull/6784

Commit:
Thread statistics addition.

API to get all system threads information and states.

SenRamakri commented on TESTS/mbed_platform/stats_thread/main.cpp:
```diff
@@ -0,0 +1,171 @@
+    t2.terminate();
+
+    count = mbed_stats_thread_get_each(stats, MAX_THREAD_STATS);
+    TEST_ASSERT_EQUAL(3, count);
```
@deepikabhavnani - You may want to check for the difference in the number of threads and not the absolute count. Because if we start running more threads as part of test infrastructure or for any other reason this test may break.

SenRamakri commented on platform/mbed_stats.c:
```diff
@@ -65,6 +65,36 @@ size_t mbed_stats_stack_get_each(mbed_stats_stack_t *stats, size_t count)
-#warning Stack statistics are currently not supported without the rtos.
+size_t mbed_stats_thread_get_each(mbed_stats_thread_t *stats, size_t count)
+{
+    MBED_ASSERT(stats != NULL);
```
Should we also assert for if count == 0, for debug builds?

deepikabhavnani replied:
In case of count 0, we will not fill up anything. But in case of NULL pointer, next operation in the code is `memset` which should not be performed on NULL pointer.

0xc0170 commented on platform/mbed_stats.h:
```diff
@@ -41,6 +47,19 @@ typedef struct {
+    uint32_t thread_stack_size;     /**< Thread Stack Size */
+    uint32_t thread_stack_space;    /**< Thread remaining stack size */
+    const char   *thread_name;      /**< Thread Object name */
+} mbed_stats_thread_t;
```
Do we need to repeat thread_ prefix ?

`mbed_stats_thread_t` says  this contains tread stats. all members are related  to the thread (if not they would be a separate object?). Thus would be thread_stats.id/state/priority oppose to thread_stats.thread_id ?

0xc0170 commented on platform/mbed_stats.h:
```diff
@@ -76,6 +95,16 @@ void mbed_stats_stack_get(mbed_stats_stack_t *stats);
+ *  @param stats    A pointer to an array of mbed_stats_thread_t structures to fill
+ *  @param count    The number of mbed_stats_thread_t structures in the provided array
+ *  @return         The number of mbed_stats_thread_t structures that have been filled,
+ *                  this is equal to the number of stacks on the system.
```
`this is equal to the number of stacks on the system` - I do not understand stacks in this  context? how many stacks we have? not threads here?

Commit:
Checking difference in thread count + review comments

Commit:
Cleaup and review comments addressed

1. Removed prefix thread_ from all elements of mbed_stats_thread_t
2. #if conditions aligned to rest of the file

Commit:
Cleanup

1. Enable watermark to get stack space information
2. Restructured code
3. Throw error if MBED_THREAD_STATS_ENABLED is not set
4. Astyle changes

cmonr commented:
/morph build

deepikabhavnani commented:
Test will fail till - MBED_ALL_STATS_ENABLED is enabled in CI testing

studavekar commented:
Will have the stats enabled in pipeline and kick off the build.

0xc0170 commented:
/morph build

cmonr commented:
/morph test

deepikabhavnani commented:
@cmonr - As per the build log,  MBED_ALL_STATS_ENABLED was not enabled. Can kill test.

deepikabhavnani commented:
Rebased to resolve conflict

deepikabhavnani commented:
@SenRamakri @0xc0170 - Please review

0xc0170 commented:
/morph build

deepikabhavnani commented:
@0xc0170 - Command used in build does not contain ALL flags enabled. Will have to re-run once @studavekar does that change.

studavekar commented:
re-triggering 

/morph build
